clients/src/main/java/org/apache/kafka/common/requests/StopReplicaRequest.java

```diff
@@ -17,75 +17,93 @@
+                       List<StopReplicaTopicState> topicStates) {
             super(ApiKeys.STOP_REPLICA, version, controllerId, controllerEpoch, brokerEpoch);
-            this.deletePartitions = deletePartitions;
-            this.partitions = partitions;
+            this.topicStates = topicStates;
         }
 
         public StopReplicaRequest build(short version) {
             StopReplicaRequestData data = new StopReplicaRequestData()
                 .setControllerId(controllerId)
                 .setControllerEpoch(controllerEpoch)
-                .setBrokerEpoch(brokerEpoch)
-                .setDeletePartitions(deletePartitions);
-
-            if (version >= 1) {
-                Map<String, List<Integer>> topicPartitionsMap = CollectionUtils.groupPartitionsByTopic(partitions);
-                List<StopReplicaTopic> topics = topicPartitionsMap.entrySet().stream().map(entry ->
-                    new StopReplicaTopic()
-                        .setName(entry.getKey())
-                        .setPartitionIndexes(entry.getValue())
-                ).collect(Collectors.toList());
+                .setBrokerEpoch(brokerEpoch);
+
+            if (version >= 3) {
+                data.setTopicStates(topicStates);
+            } else if (version >= 1) {
+                data.setDeletePartitions(deletePartitions());
+                List<StopReplicaTopicV1> topics = topicStates.stream().map(topic ->
+                    new StopReplicaTopicV1()
+                        .setName(topic.topicName())
+                        .setPartitionIndexes(topic.partitionStates().stream()
+                            .map(StopReplicaPartitionState::partitionIndex)
+                            .collect(Collectors.toList())))
+                    .collect(Collectors.toList());
                 data.setTopics(topics);
             } else {
-                List<StopReplicaPartitionV0> requestPartitions = partitions.stream().map(tp ->
-                    new StopReplicaPartitionV0()
-                        .setTopicName(tp.topic())
-                        .setPartitionIndex(tp.partition())
-                ).collect(Collectors.toList());
-                data.setUngroupedPartitions(requestPartitions);
+                data.setDeletePartitions(deletePartitions());
+                List<StopReplicaPartitionV0> partitions = topicStates.stream().flatMap(topic ->
+                    topic.partitionStates().stream().map(partition ->
+                        new StopReplicaPartitionV0()
+                            .setTopicName(topic.topicName())
+                            .setPartitionIndex(partition.partitionIndex())))
+                    .collect(Collectors.toList());
+                data.setUngroupedPartitions(partitions);
             }
 
             return new StopReplicaRequest(data, version);
         }
 
+        private boolean deletePartitions() {
```

Comment:
Might not be too big of a problem, but it would be nice to avoid this pass through all the partitions. As it is we have 1) first pass to split partitions, 2) second pass to validate split, 3) third pass to convert to the needed type. Seems like we should be able to save some work here, like perhaps moving the conversion to the caller (even though it's annoying).

Reply:
You're right. I did not realise this. I think that the best way is to move everything to the caller. I will do this.

Reply:
I have refactored this and pushed all the conversion to the caller.